File: src/main/java/earth/terrarium/tempad/client/apps/impl/SettingsApp.java
package earth.terrarium.tempad.client.apps.impl;

import earth.terrarium.tempad.api.apps.TempadApp;
import earth.terrarium.tempad.client.screens.TempadScreen;
import earth.terrarium.tempad.common.Tempad;
import earth.terrarium.tempad.common.utils.LookupLocation;
import net.minecraft.client.gui.components.WidgetSprites;
import net.minecraft.network.chat.Component;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.entity.player.Player;
import org.jetbrains.annotations.Nullable;

public class SettingsApp implements TempadApp {
    public static final SettingsApp INSTANCE = new SettingsApp();
    public static final ResourceLocation ID = new ResourceLocation(Tempad.MODID, "settings");
    public static final WidgetSprites WIDGET_SPRITES = new WidgetSprites(
        new ResourceLocation(Tempad.MODID, "apps/settings/normal"),
        new ResourceLocation(Tempad.MODID, "apps/settings/disabled"),
        new ResourceLocation(Tempad.MODID, "apps/settings/hover")
    );

    @Override
    public WidgetSprites getWidgetSprites() {
        return WIDGET_SPRITES;
    }

    @Override
    public void openOnClient(Player player, LookupLocation lookup) {

    }

    @Override
    public void openOnServer(Player player, LookupLocation lookup) {

    }
}

File: src/main/kotlin/earth/terrarium/tempad/common/apps/SettingsApp.kt
package earth.terrarium.tempad.common.apps

import com.teamresourceful.bytecodecs.base.ByteCodec
import com.teamresourceful.bytecodecs.base.`object`.ObjectByteCodec
import com.teamresourceful.resourcefullib.common.bytecodecs.ExtraByteCodecs
import earth.terrarium.tempad.api.app.TempadApp
import earth.terrarium.tempad.api.access.ItemAccessAddress
import earth.terrarium.tempad.api.locations.TempadLocations
import earth.terrarium.tempad.common.registries.ModMenus
import net.minecraft.network.chat.Component
import net.minecraft.resources.Identifier
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.entity.player.Inventory
import net.minecraft.world.entity.player.Player
import net.minecraft.world.inventory.AbstractContainerMenu
import java.util.*

data class SettingsApp(val ctx: ItemAccessAddress<*>, val isStationary: Boolean) : TempadApp<SettingsData> {
    override fun createMenu(pContainerId: Int, pPlayerInventory: Inventory, pPlayer: Player): AbstractContainerMenu {
        return ModMenus.SettingsMenu(
            pContainerId,
            pPlayerInventory,
            Optional.of(SettingsData(TempadLocations.registry.keys, isStationary, ctx))
        )
    }

    override fun getDisplayName(): Component = Component.translatable("app.tempad.settings")

    override fun createContent(player: ServerPlayer): SettingsData =
        SettingsData(TempadLocations.registry.keys, isStationary, ctx)

    override fun isEnabled(player: Player): Boolean = true
}

class SettingsData(val providers: Set<Identifier>, isStationary: Boolean, ctx: ItemAccessAddress<*>) : AppContent<SettingsData>(ctx, isStationary, codec) {
    companion object {
        val codec: ByteCodec<SettingsData> = ObjectByteCodec.create(
            ExtraByteCodecs.IDENTIFIER.setOf().fieldOf { it.providers },
            ByteCodec.BOOLEAN.fieldOf { it.isStationary },
            ItemAccessAddress.codec.fieldOf { it.ctx },
            ::SettingsData
        )
    }
}